## 2022_11_07_1319-notes

### test7_hg38_F19K16_F24B22_200k off cluster

- ```bash
  $ cd /cluster/home/t110409uhn/git/tcge-cfmedip-seq-pipeline/workflow
  $ bash bash_snakemake_mingtest7_hg38_F19K16_F24B22_200k.sh
  ...
  [Mon Nov  7 14:24:49 2022]
  rule meth_qc_quant:
  ...
  Calculating calibration curve...
  Estimating and correcting CG bias for reads mapped to the DNA negative strand...
  /usr/bin/bash: line 1:  7913 Killed                  ( Rscript --vanilla /cluster/home/t110409uhn/git/tcge-cfmedip-seq-pipeline/workflow/scripts/medips_medestrand_qsea.R A dedup_bam_umi_pe/A_dedup.bam BSgenome.Hsapiens.UCSC.hg38 True 300 /cluster/home/t110409uhn/git/tcge-cfmedip-seq-pipeline/workflow/dependencies/MeDEStrand ) 2> logs/A_medips_medestrand_qsea.log
  ```

- out of memory on short node with 8Gb mem

### test7_hg38_F19K16_F24B22_200k ON cluster, 1 core, 16G mem

- ```bash
  $ cd /cluster/home/t110409uhn/git/tcge-cfmedip-seq-pipeline/workflow
  $ sbatch sbatch_snakemake_mingtest7_hg38_F19K16_F24B22_200k_cluster.sh
  
  $ cd /cluster/projects/pughlab/projects/cfMeDIP_compare_pipelines/tcge-cfmedip-seq-pipeline-mingtest/test7_hg38_F19K16_F24B22_200k_cluster
  drwxr-xr-x 3 t110409uhn pughlab  950 Nov  7 15:45 aggregated
  drwxr-xr-x 3 t110409uhn pughlab  228 Nov  7 15:40 aggregated_spikein
  drwxr-xr-x 2 t110409uhn pughlab  941 Nov  7 15:51 autos_bfilt
  drwxr-xr-x 2 t110409uhn pughlab   31 Nov  7 14:58 barcoded_fq_pe
  -rw-r--r-- 1 t110409uhn pughlab 2944 Nov  7 14:41 config_mingtest.yaml
  drwxr-xr-x 2 t110409uhn pughlab  107 Nov  7 15:26 dedup_bam_spikein
  drwxr-xr-x 2 t110409uhn pughlab  231 Nov  7 15:08 dedup_bam_umi_pe
  drwxr-xr-x 2 t110409uhn pughlab  292 Nov  7 14:59 fastqc_pe
  drwxr-xr-x 2 t110409uhn pughlab  348 Nov  7 15:19 fragment_profile
  drwxr-xr-x 2 t110409uhn pughlab   88 Nov  7 15:11 fragment_size
  drwxr-xr-x 2 t110409uhn pughlab   88 Nov  7 15:38 fragment_size_spikein
  drwxr-xr-x 2 t110409uhn pughlab  544 Nov  7 15:43 logs
  drwxr-xr-x 2 t110409uhn pughlab 2660 Nov  7 15:49 logs_cluster
  drwxr-xr-x 2 t110409uhn pughlab  437 Nov  7 15:41 meth_qc_quant
  drwxr-xr-x 2 t110409uhn pughlab  229 Nov  7 15:29 meth_qc_quant_spikein
  drwxr-xr-x 2 t110409uhn pughlab  104 Nov  7 15:04 raw_bam
  -rw-r--r-- 1 t110409uhn pughlab   34 Nov  7 14:40 toy_sample_aggr.tsv
  -rw-r--r-- 1 t110409uhn pughlab  416 Nov  7 14:40 toy_sample_pe.tsv
  drwxr-xr-x 2 t110409uhn pughlab  102 Nov  7 15:02 trimmed_fq
  ## all done
  ## ~1.0hr run
  ```

### test8_hg38_F19K16_F24B22_50k

- ```bash
  Calculating calibration curve...
  [Mon Nov  7 12:33:19 2022]
  Error in rule meth_qc_quant:
  ## log
  ...
  Error in MEDIPS.calibrationCurve(MSet = MSet1[[i]], CSet = CSet, input = F) : 
    The dependency of coverage signals on sequence pattern (e.g. CpG) densities is different than expected. No linear model can be build, please check the calibration plot by providing the MSet object at ISet.
  Calls: MEDIPS.meth -> MEDIPS.calibrationCurve
  Execution halted
  ...
  ```

- 50k too few reads

### test9_hg38_F19K16_F24B22_100k

- ```bash
  $ cd /cluster/home/t110409uhn/git/tcge-cfmedip-seq-pipeline/workflow
  $ bash bash_snakemake_mingtest9_hg38_F19K16_F24B22_100k.sh
  ...
  Calculating calibration curve...
  [Mon Nov  7 15:12:14 2022]
  Error in rule meth_qc_quant:
  ## log
  ...
  Error in MEDIPS.calibrationCurve(MSet = MSet1[[i]], CSet = CSet, input = F) : 
    The dependency of coverage signals on sequence pattern (e.g. CpG) densities is different than expected. No linear model can be build, please check the calibration plot by providing the MSet object at ISet.
  Calls: MEDIPS.meth -> MEDIPS.calibrationCurve
  Execution halted
  ...
  ```

### others

- add resource config for each step in config? that way can see all steps too
- config file missing in github, config_template link broken
- 

<br>

## EOF